### ISSUE TITLE
Query with Contains and FirstOrDefault that worked in EF7 no longer works in EF8

### ISSUE THREAD

**Issue by ngelotte** [dotnet/efcore#32423]
When upgrading to EF8 from EF7 one of our queries will no longer generate sql but instead errors out with the following error:
```
The LINQ expression '__ids_0
    .Contains(MaterializeCollectionNavigation(
        Navigation: Client.ClientRelatedEntities,
        subquery: DbSet<ClientRelatedEntity>()
            .Where(i => EF.Property<int?>(StructuralTypeShaperExpression: 
                OurDb.Client
                ValueBufferExpression: 
                    ProjectionBindingExpression: EmptyProjectionMember
                IsNullable: False
            , "ClientId") != null && object.Equals(
                objA: (object)EF.Property<int?>(StructuralTypeShaperExpression: 
                    OurDb.Client
                    ValueBufferExpression: 
                        ProjectionBindingExpression: EmptyProjectionMember
                    IsNullable: False
                , "ClientId"), 
                objB: (object)EF.Property<int?>(i, "ClientId"))))
        .Select(t => t.SourceId)
        .FirstOrDefault())' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
```
Repro

With the following Classes as tables in the Context
```C#
public class Client
    {
        public int ClientId { get; set; }
        public string ClientName { get; set; }
        public ICollection<ClientRelatedEntity> ClientRelatedEntities { get; set; }
    }
    public class ClientRelatedEntity
    {
        public int ClientRelatedEntityId { get; set; }
        public int ClientId { get; set; }
        public Client Client { get; set; }
        public string SourceId { get; set; }
    }

OurDbContext db = new OurDbContext();
   List<string> ids = new List<string>() { "Thing1","Thing2", "Thing3" };
   var test2 = await (
       from client in db.Clients
       where ids.Contains(client.ClientRelatedEntities.Select(t => t.SourceId).FirstOrDefault())
       select client).ToListAsync();
   Console.WriteLine($"Got {test2.Count} clients");
```

### Include provider and version information

EF Core version: 8.0
Database provider: e.g. Microsoft.EntityFrameworkCore.SqlServer
Target framework: .net core 8
Operating system: Windows
IDE: Visual Studio 2022 17.8.0

**Comment by roji** [comment]
Duplicate of #32217

**Comment by roji** [comment]
Confirmed that the test code above fails with 8.0.0, but works with the release/8.0 branch. This will be released with 8.0.2.
